aptos-move/framework/aptos-stdlib/sources/cryptography/bulletproofs.move

```diff
@@ -0,0 +1,266 @@
+            &ristretto255::basepoint(), &ristretto255::hash_to_point_base(),
+            proof.bytes,
+            num_bits,
+            dst
+        )
+    }
+
+    /// Verifies a zero-knowledge range proof that the value `v` encrypted by `ct` with ElGamal public key `pubkey` satisfies $v \in [0, 2^{num_bits})$. Only works
+    /// for `num_bits` \in {8, 16, 32, 64}.
+    public fun verify_range_proof_elgamal(ct: &elgamal::Ciphertext, proof: &RangeProof, pubkey: &elgamal::Pubkey, num_bits: u64, dst: vector<u8>): bool {
+        assert!(features::bulletproofs_enabled(), E_NATIVE_FUN_NOT_AVAILABLE);
+
+        verify_range_proof_internal(
+            ristretto255::point_to_bytes(&elgamal::get_value_component_compressed(ct)),
+	    &ristretto255::basepoint(), &elgamal::get_point_from_pubkey(pubkey),
+            proof.bytes,
+            num_bits,
+            dst
+        )
+    }
+
+    /// Verifies a zero-knowledge range proof that the value `v` committed in `com` (as v * val_base + r * rand_base,
+    /// for some randomness `r`) satisfies $v \in [0, 2^{num_bits})$. Only works for `num_bits` \in {8, 16, 32, 64}.
+    public fun verify_range_proof(
+        com: &RistrettoPoint,
+        val_base: &RistrettoPoint, rand_base: &RistrettoPoint,
+        proof: &RangeProof, num_bits: u64, dst: vector<u8>): bool
+    {
+        assert!(features::bulletproofs_enabled(), E_NATIVE_FUN_NOT_AVAILABLE);
+
+        verify_range_proof_internal(
+            ristretto255::point_to_bytes(&ristretto255::point_compress(com)),
+            val_base, rand_base,
+            proof.bytes, num_bits, dst
+        )
+    }
+
+    #[test_only]
+    /// Computes a range proof for the Pedersen commitment to 'val' with 'randomness', under the default Bulletproofs
+    /// commitment key; see `pedersen::new_commitment_for_bulletproof`. Returns the said commitment too.
+    ///  Only works for `num_bits` \in {8, 16, 32, 64}.
+    public fun prove_range_pedersen(val: &Scalar, r: &Scalar, num_bits: u64, dst: vector<u8>): (RangeProof, pedersen::Commitment) {
+        let (bytes, compressed_comm) = prove_range_internal(scalar_to_bytes(val), scalar_to_bytes(r), num_bits, dst);
+        let point = ristretto255::new_compressed_point_from_bytes(compressed_comm);
+        let point = &std::option::extract(&mut point);
+
+        (
+            RangeProof { bytes },
+            pedersen::new_commitment_from_compressed(point)
+        )
+    }
+
+    //
+    // Native functions
+    //
+
+    native fun verify_range_proof_internal(
+        com: vector<u8>,
+        val_base: &RistrettoPoint,
+        rand_base: &RistrettoPoint,
+        proof: vector<u8>,
+        num_bits: u64,
+        dst: vector<u8>): bool;
+
+    #[test_only]
+    /// Returns a tuple consisting of (1) a range proof for 'val' committed with randomness 'r' under the default Bulletproofs
+    /// commitment key and (2) the commitment itself.
+    native fun prove_range_internal(
```

Comment:
Let's apply the same changes here as for `verify_range`:
 1. Single `prove_range_internal` native that takes the `val_base` and the `rand_base` as input.
 2. Three `#[test_only]` public functions that call this native:
      a) `prove_range_pedersen`
      b) `prove_range_elgamal`
      c) `prove_range`